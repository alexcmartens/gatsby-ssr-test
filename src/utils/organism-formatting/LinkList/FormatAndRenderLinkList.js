import React, { useState, useContext } from "react"
import formatTheme from "../formatTheme"
import renderPaddingOrFullBleed from "../renderPaddingOrFullBleed"
import { Link } from "gatsby"
import formatVideo from "../formatVideo"
import { LinkList } from "terra-one"
import { LayoutContext } from "../../../components/Layout/Layout"
import formatPageLinks, { updateDownloadLink } from "../formatPageLinks"

const FormatAndRenderLinkList = ({
  data,
  paddingClass,
  index,
  location,
  overrideTheme,
  nested,
  locale,
  currentSpaceId,
  products,
  siteIdentifier,
}) => {

  const [override, setOverride] = useState(false)
  const { homepageSlug } = useContext(LayoutContext)
  // if link list is nested in a component, we need to make sure there is no additional padding to fix alignment issue. We set contentWidth to false in Superhero/FullbackgroundHero/etc to represent that we should NOT have additional padding. If link list if by itself, there should be padding.
  const contentWidth = data?.contentWidth === false ? false : true

  if (!overrideTheme) {
    if (data?.colorScheme?.toLowerCase() === "light") {
      data.theme = "white"
    } else if (data?.colorScheme?.toLowerCase() === "transparent-dark-text") {
      data.theme = "transparent-dark-text"
    } else if (data?.colorScheme?.toLowerCase() === "transparent-light-text") {
      data.theme = "transparent-light-text"
    } else {
      data.theme = "dark-navy"
    }
  } else {
    data.theme = overrideTheme?.toLowerCase()
  }

  const links = formatLinks(
    data?.links,
    data?.icon,
    locale,
    override,
    setOverride,
    data,
    homepageSlug,
    currentSpaceId,
    products,
    siteIdentifier
  )
  const theme = formatTheme(data)
  return (
    <>
      {renderPaddingOrFullBleed(
        <LinkList links={links} key={index} theme={theme} />,
        contentWidth,
        true,
        true,
        nested
      )}
    </>
  )
}

export default FormatAndRenderLinkList

export const formatLinks = (
  links,
  icon,
  locale = "en",
  override,
  setOverride,
  data,
  homepageSlug,
  currentSpaceId,
  products,
  siteIdentifier
) => {
  if (!links) {
    return []
  }

  return links.map(link => {
    let productTrial
    if (link && link?.url?.__typename === "ContentfulProductTrialPage") {
      productTrial = formatPageLinks(
        link,
        locale,
        null,
        null,
        null,
        products,
        siteIdentifier
      )
    }

    const linkType = link?.url?.contentType
      ? link?.url?.contentType
      : link?.url?.internal?.type
      ? link?.url?.internal?.type
      : null

    let videoContent
    if (link?.video) {
      const videoData = formatVideo(link?.video)

      const video = videoData?.video

      const thumbnail = videoData?.thumbnail?.focalPointImage?.url
        ? videoData?.thumbnail
        : null

      videoContent = {
        transcript: videoData?.transcript,
        transcriptText: videoData?.transcriptText,
        closeTranscript: videoData?.closeTranscript,
        video,
        thumbnail,
        variant: "variant_7",
        videoProgress: data?.videoProgress,
        setVideoProgress: data?.setVideoProgress,
      }
    }

    let linkText
    let prefix = ``

    if (linkType && linkType === "article") {
      prefix = `articles/`
    } else if (linkType && linkType === "event") {
      prefix = `events/`
    } else if (linkType && linkType === "prose") {
      prefix = `prose/`
    } else if (linkType && linkType === "ContentfulNews") {
      prefix = `news/`
    }

    if (link?.text) {
      linkText = link.text
    }

    if (!link?.text && !productTrial) {
      linkText = link?.url?.text
    }

    if (productTrial) {
      linkText = productTrial?.text
    }
    const url = link?.internalLink ? link?.internalLink : link?.url?.url

    const destination =
      productTrial && productTrial?.url
        ? productTrial?.url
        : link?.url?.realLink?.url
        ? null
        : link?.url?.url?.startsWith("http")
        ? url
        : url === homepageSlug
        ? `/${locale?.toLowerCase()}/${prefix}`
        : `/${locale?.toLowerCase()}/${prefix}${url}`

    return {
      linkText,
      destination: destination?.includes("undefined") ? null : destination,
      icon: icon ? icon : null,
      external: destination?.startsWith("http") ? true : false,
      Link: destination?.startsWith("http") ? null : Link,
      video: videoContent,
      override,
      setOverride,
      onClick:
        link?.url?.type === "ContentfulDownloadLink"
          ? () =>
              window.open(
                updateDownloadLink(link?.url?.realLink?.url, currentSpaceId),
                "_blank"
              )
          : null,
    }
  })
}
