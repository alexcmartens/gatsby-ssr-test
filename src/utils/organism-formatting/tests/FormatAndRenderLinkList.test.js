import React from "react"
import renderer from "react-test-renderer"
import FormatAndRenderLinkList, {
  formatLinks,
} from "../LinkList/FormatAndRenderLinkList"
import { Link } from "gatsby"

describe("FormatAndRenderLinkList", () => {
  it("renders correctly", () => {
    const links = [
      { item: "item1" },
      { item: "item2" },
      { item: "item3" },
      { item: "item4" },
    ]

    const colorScheme = "light"

    const data = { links, colorScheme }

    const tree = renderer
      .create(<FormatAndRenderLinkList data={data} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe("formatLinks", () => {
  it("should not break if no links are passed in", () => {
    const result = formatLinks()
    expect(result).toEqual([])
  })

  it("should format links as expected", () => {
    const links = [
      {
        text: "text here",
        url: {
          url: "http://www.trimble.com",
          text: "Trimble Section",
        },
      },
      {
        text: null,
        url: {
          url: "http://www.trimble.com",
          text: "Trimble Section",
        },
      },
      {
        text: "text",
        url: {
          url: "page-base-test",
          text: "Page Base Test",
        },
      },
      {
        text: null,
        url: {
          url: "pasta-page",
          text: "Pasta Puppy",
        },
      },
    ]

    const result = formatLinks(links, true, "en")

    const expected = [
      {
        linkText: "text here",
        destination: "http://www.trimble.com",
        icon: true,
        external: true,
        Link: null,
        onClick: null,
        override: undefined,
        setOverride: undefined,
        video: undefined,
      },
      {
        linkText: "Trimble Section",
        destination: "http://www.trimble.com",
        icon: true,
        external: true,
        Link: null,
        onClick: null,
        override: undefined,
        setOverride: undefined,
        video: undefined,
      },
      {
        linkText: "text",
        destination: "/en/page-base-test",
        icon: true,
        external: false,
        Link: Link,
        onClick: null,
        override: undefined,
        setOverride: undefined,
        video: undefined,
      },
      {
        linkText: "Pasta Puppy",
        destination: "/en/pasta-page",
        icon: true,
        Link: Link,
        external: false,
        onClick: null,
        override: undefined,
        setOverride: undefined,
        video: undefined,
      },
    ]
    expect(result).toEqual(expected)
  })
})
