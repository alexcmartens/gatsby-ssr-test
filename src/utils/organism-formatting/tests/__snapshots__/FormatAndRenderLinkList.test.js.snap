// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormatAndRenderLinkList renders correctly 1`] = `
<section
  className="default-layout   mockconstructor
      default-layout--no-bottom-margin"
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "marginBottom": "",
    }
  }
>
  <section
    className="default-layout__grid 
          
        "
    style={
      Object {
        "backgroundColor": "",
        "marginBottom": "",
      }
    }
  >
    <div
      className="default-layout__children "
    >
      <div
        links={
          Array [
            Object {
              "Link": [MockFunction],
              "destination": "/en/undefined",
              "external": false,
              "icon": null,
              "linkText": undefined,
              "onClick": null,
              "override": false,
              "setOverride": [Function],
              "video": undefined,
            },
            Object {
              "Link": [MockFunction],
              "destination": "/en/undefined",
              "external": false,
              "icon": null,
              "linkText": undefined,
              "onClick": null,
              "override": false,
              "setOverride": [Function],
              "video": undefined,
            },
            Object {
              "Link": [MockFunction],
              "destination": "/en/undefined",
              "external": false,
              "icon": null,
              "linkText": undefined,
              "onClick": null,
              "override": false,
              "setOverride": [Function],
              "video": undefined,
            },
            Object {
              "Link": [MockFunction],
              "destination": "/en/undefined",
              "external": false,
              "icon": null,
              "linkText": undefined,
              "onClick": null,
              "override": false,
              "setOverride": [Function],
              "video": undefined,
            },
          ]
        }
        theme={
          Object {
            "accordionIconColor": "#0063A3",
            "backgroundColor": "#FFFFFF",
            "color": "#252A2E",
            "ctaOne": "primary--1",
            "ctaTwo": "primary--2",
            "darkTheme": false,
            "iconColor": "#252A2E",
            "theme": "white",
          }
        }
      />
    </div>
  </section>
</section>
`;
